sip/sip-001-burn-election.md

```diff
@@ -0,0 +1,251 @@
+transactions are merely ignored by the network.
+
+# Fork Selection
+
+Fork selection in the Stacks blockchain requires a metric to determine
+which chain, between two candidates, is the "heaviest" or "longest"
+chain. Use Proof-of-Burn as the security method for the blockchain
+implies a direct metric: the total sum of burns in the election blocks
+for the candidate chain. Even with this metric, there's a design choice--
+
+**How do we count the total burn for a given Stacks block _N_?**
+
+There are two alternatives:
+
+1. The total amount of burn required for _election_ of block _N_.
+2. The total amount of burn which _confirms_ block _N_.
+
+Option 1 would allow the "heaviest chain" to be known at the time that
+the block _N_ begins. Option 2, however, is much more resilient to
+malicious or greedy leaders--- if their blocks aren't well broadcasted
+and announced, they will not be confirmed and therefore won't have
+competitive chain weights when comparing them against alternative forks.
+
+## Weighting functions for chain length
+
+The naive chain length metric (i.e., summing over total amount burn)
+can be modified with weighting functions to encourage different
+difficulties for different kinds of forks. In particular, in order to
+discourage _deep_ forks, but still tolerate shallow forks, the chain
+length metric can use exponential weighting based on burn-chain depth,
+such that the "chain weight" applied to a given burn transaction is
+equal to:
+
+```
+O(b * 2^d)
+
+b := burn amount
+d := depth in the burn chain
+```
+
+This would strongly favor deeper confirmations, making it very
+difficult (though, in extreme circumstances, possible) to perform deep
+reorganizations.
+
+## Burn difficulty windowing
+
+In order to prevent massive swings in the chain weights, we can employ
+a technique of difficulty windowing. Namely, the total "heaviness" of
+blocks in the Stacks chain is only allowed to increase at a maximum
+rate per epoch (e.g., 5%). This prevents large spenders from trivially
+creating forks by spending large sums of cryptocurrency
+reserves. Similar to exponential weighting of block weights, this
+modification increases the difficulty of forking (or, strictly places
+a limit on how many _epochs_ a fork of a given depth must build across),
+which trades resilience towards bad leaders for chain stability.
+
+# Open Questions
+
+The above discussion outlined a few major open questions:
+
+1. How do we set the _election window?_ This is the time between when
+   an election is finalized, and the leader begins broadcasting.
+
+2. How do we account for burns in chain-selection? Does the block
+   which the burn _elected_ receive the weight? Or does the block
+   which the burn _confirmed_?
+   
+3. How do we weight burns over time? Does an exponential weighting
+   function achieve desired properties (and what is the trade-off we
+   are making here?) How should we tune these parameters?
+
```

Comment:
The longest valid chain wins, where "longest" means "has the most blocks."  Over long enough time scales, this will also be the chain that has the most burns.

The problem we are trying to solve is to decide which of a set of valid chains is the "best."  The behavior we want is for the "best" chain to be the chain that an overwhelming majority of miners compete to produce blocks on -- we want miners to reap the rewards of making progress on the already-best chain whenever possible.  We achieve this end by rewarding miners for building off of the chain with the most blocks, since blocks are a proxy for Stacks economic activity.

An alternative is to consider a chain to be "best" only if it is valid and has the most cumulative burns.  We don't encourage miners to build off of the most widely-used chain by allowing anyone to make a large burn and produce a shorter chain that is considered "better" while invalidating a lot of already-confirmed blocks.  Such chains should not be considered "better" precisely because doing so would enable this unintuitive behavior, and would allow malicious whales to reduce the system's liveness and hurt honest miners.

Nevertheless, the longest valid chain is also guaranteed to be the chain with the most burns over time, since an honest majority coalition of miners will all compete to be selected to be the next minder by burning higher BTC than their peers.  The side-effect of competing to mine on the longest chain is that the sum of the burns on that chain will also (over time) exceed the sums of burns on minority forks.

The reason we can get away with only considering chain length for quality is because PoB miners enjoy two advantages over PoW miners:  access to a global clock (via the PoW chain's block arrivals), and an arbitrarily-low barrier to entry for participating in the protocol.  In PoW chains, the act of mining serves two purposes:  to start new epochs, and to enhance chain quality.  As such, a PoW chain's fork-choice rule considers the chain with the most valid proof-of-work to be the best chain.

In PoB chains, the act of mining only enhances chain quality, since the mechanism that starts and ends epochs is outside the PoB miners' control.  This means that chain quality is not directly dependent on the amount of tokens burned; it's only dependent on the number of blocks produced on the same fork.  The only way for a PoB minority fork to overtake the majority fork is for the majority of miners (by burn) to produce blocks on the minority fork until its length exceeds the majority fork.  Since miners are competing in each epoch to be selected as the next miner (meaning that they burn a non-zero amount of tokens), over time the new majority fork will have more burns than the old majority fork.

It's important to remember that using chain length as the quality metric still allows the system to tolerate deep forks (such as to recover from a catastrophic network crash).  However, this enables two nice properties of chain reorganizations: (1) anyone can "see a reorg coming" because all blocks produced are anchored to the same history on the burn chain, and (2) deep forks take a long time to carry out.  These are particularly useful properties for a young PoB chain, because it increases the barrier to entry for a powerful malicious miner who can afford to out-burn a fledgling chain's miners.  I think this will prove useful for app-chains, for example.